api.go

```diff
@@ -0,0 +1,219 @@
+    copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+
+*/
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/http/httptrace"
+	"net/textproto"
+	"net/url"
+	"regexp"
+	"strings"
+	"time"
+)
+
+var placeholderRegex = regexp.MustCompile(`{{ *[^}}]* *}}`)
+
+// RunSimpleTest starts an HTTP/HTTPS API test of a site within crabby.  It uses Go's built-in net/http client.
+func RunApiTest(ctx context.Context, j Job, storage *Storage, client *http.Client) {
+	responses := map[string]json.RawMessage{}
+	for _, s := range j.Steps {
+		runApiTestStep(ctx, s, storage, client, responses)
+	}
+}
+
+func runApiTestStep(ctx context.Context, j JobStep, storage *Storage, client *http.Client, responses map[string]json.RawMessage) {
+	var method = strings.ToUpper(j.Method)
+	if method == "" {
+		method = http.MethodGet
+	}
+
+	body, err := replacePlaceholders(j.Body, responses)
+	if err != nil {
+		log.Printf("unable to substitute body variables in body: %v", err)
+		return
+	}
+
+	req, err := http.NewRequest(method, j.URL, strings.NewReader(body))
+	if err != nil {
+		log.Printf("unable to create request: %v", err)
+		return
+	}
+
+	if err := addHeaders(req, j, responses); err != nil {
+		log.Printf("unable to process headers: %v", err)
+		return
+	}
+
+	var t0, t1, t2, t3, t4 time.Time
+
+	trace := &httptrace.ClientTrace{
+		DNSStart: func(_ httptrace.DNSStartInfo) { t0 = time.Now() },
+		DNSDone:  func(_ httptrace.DNSDoneInfo) { t1 = time.Now() },
+		ConnectStart: func(_, _ string) {
+			if t1.IsZero() {
+				// connecting to IP
+				t1 = time.Now()
+			}
+		},
+		ConnectDone: func(net, addr string, err error) {
+			if err != nil {
+				log.Printf("unable to connect to host %v: %v", addr, err)
+			}
+			t2 = time.Now()
+		},
+		GotConn:              func(_ httptrace.GotConnInfo) { t3 = time.Now() },
+		GotFirstResponseByte: func() { t4 = time.Now() },
+	}
+
+	// We'll use our Context in this request in case we have to shut down midstream
+	req = req.WithContext(httptrace.WithClientTrace(ctx, trace))
+
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Println("Failed to read response:", err)
+		return
+	}
+
+	// Send our server response code as an event
+	storage.EventDistributor <- j.makeEvent(resp.StatusCode)
+
+	// unmarshal response and save into map
+	if responses[j.Name], err = ioutil.ReadAll(resp.Body); err != nil {
+		log.Println("WARNING!", err)
```

Comment:
```suggestion
		log.Println("warning: could not read response body:", err)
```

Reply:
how do you make change suggestions?

Reply:
use 

````
```suggestion
your suggestion here
```
````

```suggestion
your suggestion here
```